lib/facter/custom_facts/core/execution/base.rb

```diff
@@ -100,18 +100,22 @@ def execute_command(command, on_fail, logger = nil, time_limit = nil)
             @log.debug("Executing command: #{command}")
             out, stderr = Open3.popen3(opts, command.to_s) do |_, stdout, stderr, wait_thr|
               pid = wait_thr.pid
-              output = +''
-              err = +''
+              stdout_messages = +''
+              stderr_messages = +''
+              out_reader = Thread.new { stdout.read }
+              err_reader = Thread.new { stderr.read }
               begin
                 Timeout.timeout(time_limit) do
-                  output << stdout.read
-                  err << stderr.read
+                  stdout_messages << out_reader.value
+                  stderr_messages << err_reader.value
                 end
               rescue Timeout::Error
                 @log.debug("Timeout encounter after #{time_limit}s, killing process with pid: #{pid}")
                 Process.kill('KILL', pid)
+                out_reader.kill
+                err_reader.kill
```

Comment:
Can `Process.kill` raise such as if the child process no longer exists? If so then kill is never called on the readers.

Reply:
i wrapped the threads kill in an `ensure` block